src/store/contact-groups/index.ts

```diff
@@ -4,48 +4,119 @@
-			storedContactGroups: get().storedContactGroups.map((cg) =>
-				cg.id === contactGroup.id ? contactGroup : cg
-			)
-		})),
-	setStoredOffset: (offset): void => set(() => ({ storedOffset: offset })),
-	addStoredContactGroups: (contactGroups): void =>
-		set(() => ({ storedContactGroups: [...(get().storedContactGroups ?? []), ...contactGroups] })),
-	emptyStoredContactGroups: (): void => set(() => ({ storedContactGroups: [] })),
-	removeStoredContactGroup: (contactGroupId: string): void => {
-		const idx = get().storedContactGroups.findIndex(
+	orderedContactGroups: [],
+	unorderedContactGroups: [],
+	offset: 0,
+	updateContactGroup: (contactGroup): void => {
+		get().removeContactGroup(contactGroup.id);
+		get().addContactGroupInSortedPosition(contactGroup);
+	},
+	setOffset: (offset): void => set(() => ({ offset })),
+	addContactGroups: (contactGroups): void => {
+		const { orderedContactGroups, unorderedContactGroups } = get();
+
+		if (unorderedContactGroups.length > 0) {
+			const unorderedResult = differenceBy(unorderedContactGroups, contactGroups, (cg) => cg.id);
+			set(() => ({
+				orderedContactGroups: [...(orderedContactGroups ?? []), ...contactGroups],
+				unorderedContactGroups: unorderedResult
+			}));
+		} else {
+			set(() => ({
+				orderedContactGroups: [...(orderedContactGroups ?? []), ...contactGroups]
+			}));
+		}
+	},
+	emptyContactGroups: (): void =>
+		set(() => ({ orderedContactGroups: [], unorderedContactGroups: [] })),
+	removeContactGroup: (contactGroupId: string): void => {
+		const { orderedContactGroups, unorderedContactGroups, offset } = get();
+		const idx = orderedContactGroups.findIndex(
 			(contactGroup) => contactGroup.id === contactGroupId
 		);
 		if (idx >= 0) {
 			set(() => ({
 				// TODO replace with Array toSpliced when will be available
-				storedContactGroups: get().storedContactGroups.filter(
+				orderedContactGroups: orderedContactGroups.filter(
 					(contactGroup) => contactGroup.id !== contactGroupId
 				),
-				storedOffset: get().storedOffset - 1
+				offset: offset - 1
+			}));
+		} else {
+			const uIdx = unorderedContactGroups.findIndex(
+				(contactGroup) => contactGroup.id === contactGroupId
+			);
+			if (uIdx >= 0) {
+				set(() => ({
+					// TODO replace with Array toSpliced when will be available
+					unorderedContactGroups: unorderedContactGroups.filter(
+						(contactGroup) => contactGroup.id !== contactGroupId
+					)
+				}));
+			} else {
+				throw new Error('Contact group not found');
+			}
+		}
+	},
+	addContactGroupInSortedPosition: (newContactGroup: ContactGroup): void => {
+		const { orderedContactGroups, unorderedContactGroups, offset } = get();
+
+		const idx = findIndex(
+			orderedContactGroups,
+			(contactGroup) => compareContactGroupName(newContactGroup.title, contactGroup.title) < 0
+		);
+
+		if (idx < orderedContactGroups.length && idx >= 0) {
+			set(() => ({
+				orderedContactGroups: [...orderedContactGroups].splice(idx, 0, newContactGroup),
```

Comment:
This is not working :smile:. Splice returns the array of removed elements, and thus it is emptying the list. We need to do the splice in another line or use a different function